kong/db/dao/plugins/go.lua

```diff
@@ -0,0 +1,487 @@
+    int close(int fd);
+  ]])
+
+  local ffi_sock = ffi.typeof("ffi_sock")
+
+  pcall(ffi.metatype, ffi_sock, {
+    __index = {
+      send = function (self, s)
+        local p = ffi.cast('const uint8_t *', s)
+        local sent_bytes = 0
+        while sent_bytes < #s do
+          local rc = C.write(self.fd, p+sent_bytes, #s-sent_bytes)
+          if rc < 0 then
+            local errno = ffi.errno()
+            return nil, "error writing to socket: " .. ffi.string(C.strerror(errno))
+          end
+          sent_bytes = sent_bytes + rc
+        end
+        return sent_bytes
+      end,
+
+      receiveany = function (self, n)
+        local buf = ffi.new('uint8_t[?]', n)
+        if buf == nil then
+          return nil, "can't allocate buffer."
+        end
+
+        local rc = C.read(self.fd, buf, n)
+        if rc < 0 then
+          return nil, "error reading"
+        end
+        return ffi.string(buf, rc)
+      end,
+
+      setkeepalive = function (self)
+        if self.fd ~= -1 then
+          C.close(self.fd)
+          self.fd = -1
+        end
+      end,
+    },
+
+    __gc = function (self)
+      self:setkeepalive()
+    end,
+  })
+
+  local function un_addr(path)
+    local conaddr = ffi.new('struct sockaddr_un')
+    if conaddr == nil then
+      return nil, "can't allocate sockaddr_un"
+    end
+
+    conaddr.sun_family = C.AF_UNIX
+
+    local len = #path + 1
+    if len > 107 then
+      len = 107
+    end
+    ffi.copy(conaddr.sun_path, path, len)
+    return conaddr
+  end
+
+  local too_early = {
+    init = true,
+    init_worker = true,
+    set = true,
+    header_filter = true,
+    body_filter = true,
+--     log = true,
+  }
+
+  function get_connection(socket_path)
+    if not too_early[ngx.get_phase()] then
+--       ngx.log(ngx.DEBUG, "not early anymore, use a cosocket!")
+      return ngx.socket.connect("unix:" .. socket_path)
+    end
+
+    local fd = C.socket(C.AF_UNIX, C.SOCK_STREAM, 0)
+    if fd < 0 then
+      return nil, "can't create socket"
+    end
+
+    local conaddr = un_addr(socket_path)
+    local res = C.connect(fd, conaddr, ffi.sizeof(conaddr))
+    if res < 0 then
+      C.close(fd)
+      local errno = ffi.errno()
+      return nil, "connect failure: " .. ffi.string(C.strerror(errno))
+    end
+
+    return ffi_sock(fd)
+  end
+end
+go.get_connection = get_connection
+
+
+-- This is the MessagePack-RPC implementation
+local rpc_call
+local set_plugin_dir
+do
+  local msg_id = 0
+
+  local notifications = {}
+
+  local current_plugin_dir
+
+  do
+    local pluginserver_pid
+    function notifications.serverPid(n)
+      n = tonumber(n)
+      if pluginserver_pid and n ~= pluginserver_pid then
+        reset_instances()
+        current_plugin_dir = nil
+      end
+
+      pluginserver_pid = n
+    end
+  end
+
+  function rpc_call(method, ...)
+    msg_id = msg_id + 1
+    local my_msg_id = msg_id
+
+    local c = assert(get_connection(kong.configuration.pluginserver_socket))
```

Comment:
note that if, as I suggested, `pluginserver_socket` is no longer a property, then this cannot be an `assert` anymore. If one doesn't have the socket set up, then failure here should be handled instead.

Reply:
(The above comment is assuming that the paths leading here would be no longer guarded by `if pluginserver_socket ~= "off"`, but perhaps they could still be guarded if the first connection attempt at `kong.init` sets a boolean instead? Handling the error and bubbling up might still be more resilient, keeping process management in mind.)